scalding-core/src/main/scala/com/twitter/scalding/typed/TypedPipe.scala

```diff
@@ -1131,6 +1133,47 @@ class MappablePipeJoinEnrichment[T](pipe: TypedPipe[T]) {
 }
 
 /**
+ * This class is to enrich the forceToDisk operation on TypedPipes and allow
+ * us to potentially optimize these operations when they're chained together
+ * redundantly.
+ */
+case class ForceToDiskTypedPipe[T](typedPipe: TypedPipe[T]) extends TypedPipe[T] {
+  import Dsl._
+
+  override def forceToDisk: TypedPipe[T] = this
+
+  override def asPipe[U >: T](fieldNames: Fields)(implicit flowDef: FlowDef, mode: Mode, setter: TupleSetter[U]): Pipe = {
+    val pipe = typedPipe.toPipe[U](fieldNames)(flowDef, mode, setter)
+    pipe.forceToDisk
+  }
+
+  override def cross[U](tiny: TypedPipe[U]): TypedPipe[(T, U)] =
+    forceBefore(typedPipe).cross(tiny)
+
+  override def flatMap[U](f: T => TraversableOnce[U]): TypedPipe[U] =
+    forceBefore(typedPipe).flatMap(f)
```

Comment:
actually, I think we don't want forceBefore at all. Don't we always want: `physicalForceToDisk.flatMap(f)`?

If the user has a `groupBy`, that is materialized, but `groupBy.map(f)` will glue the `f` into the reducer. That is not what they mean when they do a `.forceToDisk`.

I'm starting to wonder if this whole `ForceToDisk` class is a red-herring.